_posts/2022-09-14-utilizing-state-machines-in-front-end-development.md

```diff
@@ -0,0 +1,543 @@
+              cond: "hasResults",
+              target: "showResults",
+              actions: assign({
+                results: (context, event) => event.results,
+              }),
+            },
+            {
+              cond: "hasNoResults",
+              target: "noResults",
+              actions: assign({
+                results: (context, event) => [],
+              }),
+            },
+          ],
+          ERROR: "failure",
+        },
+      },
+      failure: {
+        type: "final",
+      },
+      noResults: {
+        type: "final",
+      },
+      showResults: {
+        type: "final",
+      },
+    },
+  },
+  {
+    guards: {
+      hasResults: (context, event) => event.results.length > 0,
+      hasNoResults: (context, event) => event.results.length === 0,
+    },
+  }
+);
+```
+
+On the component, we may now remove the `useState` hooks for the `keyword` and the `results` and access the machine's context using `state.context`. We were already sending the results as part of the `SUCCESS` event, so no changes are required there. On the input field's `onChange` handler we can straight up call the `send` function and send the `SET_KEYWORD` event with the field value as data.
+
+```jsx
+import { useMachine } from "@xstate/react";
+import { searchFormMachine } from "./searchFormMachine.js";
+
+const SearchForm = () => {
+  const [state, send] = useMachine(searchFormMachine);
+
+  const search = (event) => {
+    event.preventDefault();
+
+    send("SEARCH");
+
+    fetchResults(state.context.keyword)
+      .then((results) => {
+        send({ type: "SUCCESS", data: results });
+      })
+      .catch((error) => send("ERROR"));
+  };
+
+  return (
+    <div>
+      <form onSubmit={search}>
+        <input
+          onChange={(event) =>
+            send({ type: "SET_KEYWORD", data: event.target.value })
+          }
+        />
+        <button type="submit" disabled={state.value === "loading"}>
+          Search
+        </button>
+      </form>
+
+      {state.value === "initial" && <p>Hi! Enter keyword please.</p>}
+      {state.value === "loading" && <p>Searching...</p>}
+      {state.value === "failure" && <p>Oh no!</p>}
+      {state.value === "noResults" && <p>Sorry! No results.</p>}
+      {state.value === "showResults" &&
+        state.context.results.map((result) => <p>{result}</p>)}
+    </div>
+  );
+};
+```
+
+Cool, now the component's whole state is handled by XState. We have successfully moved the business logic out from the view layer handled by React. We could even move the fetching to be a part of the machine (maybe a reusable sub-machine?).
+
+And by the way, our machine looks like this now:
+
+![A visual presentation of a state machine representing a search form on a web app](/img/utilizing-state-machines-in-front-end-development/SearchFormWithXStateV2.png)
+
+The `SET_KEYWORD` event assigns the data but doesn't trigger a transition while `SUCCESS` assigns and transitions at the same time.
```

Comment:
Should it transition to a state, where the search button is enabled because we have entered more than 3 letters to the search word input?

Reply:
I'd rather keep the code simple and in sync with the non-xstate version, but I'll add this as a suggestion to the section before the image!